JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/timeseries/TimeSeriesSparql.java

```diff
@@ -175,114 +302,277 @@ protected void initTS(String timeSeriesIRI, List<String> dataIRI, String dbURL,
+		}
+
+		return durationIRI;
+
+	}
+
+	/**
+	 * Check whether given duration IRI is attached to an Average time series
+	 * @param durationIRI duration IRI provided as string
+	 * @return True if durationIRI exists and is attached to an Average time series, false otherwise
+	 */
+	private boolean checkDurationHasAverageTimeSeries(String durationIRI) {
+		String query = String.format("ask {?a <%s> <%s>}", (ns_ontology + "hasAveragingPeriod"), durationIRI);
+		kbClient.setQuery(query);
+		return kbClient.executeQuery().getJSONObject(0).getBoolean("ASK");
+	}
+
+	/**
+	 * Check whether a particular Average time series (i.e. tsIRI) exists
+	 * @param timeSeriesIRI AverageTimeSeries IRI provided as string
+	 * @return True if an average time series instance with the tsIRI exists, false otherwise
+	 */
+	private boolean checkAverageTimeSeriesExists(String timeSeriesIRI) {
+		String query = String.format("ask {<%s> a <%s>}", timeSeriesIRI, (ns_ontology + "AverageTimeSeries"));
+		kbClient.setQuery(query);
+		return kbClient.executeQuery().getJSONObject(0).getBoolean("ASK");
+	}
+
+	/**
+	 * Check whether a particular Instantaneous time series (i.e. tsIRI) exists
+	 * @param timeSeriesIRI InstantaneousTimeSeries IRI provided as string
+	 * @return True if an Instantaneous time series instance with the tsIRI exists, false otherwise
+	 */
+	private boolean checkInstantaneousTimeSeriesExists(String timeSeriesIRI) {
+		String query = String.format("ask {<%s> a <%s>}", timeSeriesIRI, (ns_ontology + "InstantaneousTimeSeries"));
+		kbClient.setQuery(query);
+		return kbClient.executeQuery().getJSONObject(0).getBoolean("ASK");
+	}
+
+	/**
+	 * Check whether a particular CumulativeTotal time series (i.e. tsIRI) exists
+	 * @param timeSeriesIRI CumulativeTotalTimeSeries IRI provided as string
+	 * @return True if a CumulativeTotal time series instance with the tsIRI exists, false otherwise
+	 */
+	private boolean checkCumulativeTotalTimeSeriesExists(String timeSeriesIRI) {
+		String query = String.format("ask {<%s> a <%s>}", timeSeriesIRI, (ns_ontology + "CumulativeTotalTimeSeries"));
+		kbClient.setQuery(query);
+		return kbClient.executeQuery().getJSONObject(0).getBoolean("ASK");
+	}
+
+	/**
+	 * Check whether a particular StepwiseCumulative time series (i.e. tsIRI) exists
+	 * @param timeSeriesIRI StepwiseCumulativeTimeSeries IRI provided as string
+	 * @return True if a StepwiseCumulative time series instance with the tsIRI exists, false otherwise
+	 */
+	private boolean checkStepwiseCumulativeTimeSeriesExists(String timeSeriesIRI) {
+		String query = String.format("ask {<%s> a <%s>}", timeSeriesIRI, (ns_ontology + "StepwiseCumulativeTimeSeries"));
+		kbClient.setQuery(query);
+		return kbClient.executeQuery().getJSONObject(0).getBoolean("ASK");
+	}
+
+	protected void bulkInitTS(List<String> tsIRIs, List<List<String>> dataIRIs, String rdbURL, List<String> timeUnit, List<String> type, List<Duration> durations, List<ChronoUnit> units) {
```

Comment:
This method should only interact with the triple-store once (one sparql update), you are sending lots of queries here (e.g. checkTimeSeriesExist) which will slowthings down